content/docs/engine-dev/tools/project-manager/source-code-map.md

```diff
@@ -0,0 +1,110 @@
+
+
+### Application
+
+Inherits from `AzFramework::Application`.
+
+The functions of the Application class are exclusively run by `main.cpp`.
+
+The two primary functions to keep track of are `Init` and `Run`. [`TearDown`](https://github.com/o3de/o3de/blob/development/Code/Tools/ProjectManager/Source/Application.cpp#L260-L272) is invoked at `Application`'s deconstruction.
+
+
+##### [`Application::Init`](https://github.com/o3de/o3de/blob/development/Code/Tools/ProjectManager/Source/Application.cpp#L34)
+* Sets various `QApplication` and `QCoreApplication` attributes and parameters
+* Registers components for system logging, and invokes [`InitLog`](https://github.com/o3de/o3de/blob/development/Code/Tools/ProjectManager/Source/Application.cpp#L152)
+* Creates new instance of `QApplication`, feeding in command line arguments
+* Sets up PythonBindings
+* Calls [`RegisterEngine`](https://github.com/o3de/o3de/blob/development/Code/Tools/ProjectManager/Source/Application.cpp#L181) to setup the O3DE engine that Project Manager ships with
+* Gets connection to CommandLine, and parses for screen and projectPath parameters
+* Creates ProjectManagerWindow
+
+##### [`Application::Run`](https://github.com/o3de/o3de/blob/development/Code/Tools/ProjectManager/Source/Application.cpp#L274)
+* Sets up Qt GUI stylesheets
+* Sets up Window decoration and MainWindow Geometry
+* Shows the MainWindow
+* Run the QApplication's main loop
+
+[Back to Top](#map)
+
+
+
+### ProjectManagerWindow
+This inherits all primary functionality from `QMainWindow`.
+
+All logic unique to the Project Manager occurs at the constructor, which is summarized as follows:
+* Fetch Engine Info to set Window Title with correct engine version information
+* Instantiate the `DownloadController`
+* Instantiate the `ScreensCtrl` for screen management
+* Construct a list of relevant `ProjectManagerScreen` enums, and use `ScreensCtrl` to construct each screen widget for Project Manager's GUI
+* Set instance of `ScreensCtrl` as central widget of the window
+* Retrieve passed in arguments for `startScreen` and `projectPath`.
+    * If `startScreen` is set, first force load the main Projects screen, then transition to the desired screen. If not set, the Projects screen is the default.
+    * If `projectPath` is set, notify the `ScreensCtrl` instance of the currently highlighted project path.
+
+[Back to Top](#map)
+
+
+
+### ScreensCtrl
+ScreensCtrl is the central widget which stores all GUI information for Project Manager. 
+
+Screens are typically accessed using `m_screenMap`. All screens are stored in `m_screenStack`. The three primary tabs of Project Manager (Project, Gem, and Engine), are stored in `m_tabWidget`. Traversal history through the screens of ProjectManager can be tracked via `m_screenVisitOrder`.
+
+All functions are summarized as follows:
```

Comment:
This is good information, but I'm going to play devil's advocate here and raise a few challenge questions:

1. Shouldn't all of these comments be in the source code itself? The level of commenting in the source in each function header and within the function body appears rather sparse. It would seem more useful to me to have some of these comments in the code itself, where it's more likely to be updated as implementation changes.
2. Is it worth documenting the function implementation in this much detail here? Two challenges this presents:

- As the source changes, the line numbers will occasionally change and invalidate the links (as mentioned elsewhere in this review).
- Any changes to the implementation of a function would require a change to this documentation, and that is going to present a maintenance challenge for the most detailed part of this documentation to remain correct and up to date.

I'm wondering how to maximize the usefulness of this kind of documentation and minimize the maintenance overhead. Would it be better to limit the focus of source code map content to just an overview of each class and function, followed in some cases by a discussion of how it fits into the overall architecture, including how it's used, and when it plays an important role in various scenarios; and then move the detailed function implementation explanations into the source as comments?

Reply:
For the current state of notes, I completely agree with you that these are basically the same value as in-code comments. I may consider going back to the source code and add these comments in later.

On the other hand, in terms of comprehension, I feel having these comments in the same place as code introduces cognitive burden on code reading, which I've experienced firsthand (and usually goes unsaid by developers). The coding guidelines in the o3de repo also states they would prefer no comments if the code can be made self-documenting instead (which is a work in progress right now).

In principle, having an explanation organized in an external place can make it easier to dive into the code directly and see what is going on, because the text density is much less. I may need to improve the text here so it's brief like a summary. If you've got suggestions on what parts can be pruned, that would be great to hear.

That being said, I feel like the current state of notes would be dramatically improved once I introduce diagrams to better describe the data structures involved. However that will take some time, since I need to make sure they are accurate in what they portray. These notes are there as a temporary stage to that point.

Reply:
also, per an earlier conversation, if the documentation gets mapped to a particular commit and stated at the top, we no longer have to fear the code making the documentation incorrect/stale, since it's effectively fixed in place.

This will mean however it will not be up-to-date as soon as another commit touches the code, which is a definite disadvantage. But I think it helps us in maintaining clarity in the long run, and we can choose to prioritize updating only for major commits which dramatically change the state of the code.